File: Homework1/Hw1/Parser.cs
namespace Hw1;

public static class Parser
{
    public static void ParseCalcArguments(string[] args, 
        out double val1, 
        out CalculatorOperation operation, 
        out double val2)
    {
        if (!IsArgLengthSupported(args))
            throw new ArgumentException("Wrong amount of arguments");
        
        if (!Double.TryParse(args[0], out val1))
            throw new ArgumentException("Values must be double");
        operation = ParseOperation(args[1]);
        if (!Double.TryParse(args[2], out val2))
            throw new ArgumentException("Values must be double");

        switch (args[1])
        {
            case "+":
                operation = CalculatorOperation.Plus;
                break;
            case "-":
                operation = CalculatorOperation.Minus;
                break;
            case "*":
                operation = CalculatorOperation.Multiply;
                break;
            case "/":
                operation = CalculatorOperation.Divide;
                break;
        }
    }

    private static bool IsArgLengthSupported(string[] args) => args.Length == 3;

    private static CalculatorOperation ParseOperation(string arg) =>
        arg switch
        {
            "+" => CalculatorOperation.Plus,
            "-" => CalculatorOperation.Minus,
            "*" => CalculatorOperation.Multiply,
            "/" => CalculatorOperation.Divide,
            _ => throw new InvalidOperationException("Wrong operation"),
        };
}

File: Homework4/Hw4/Parser.fs
module Hw4.Parser

open System
open Hw4.Calculator


type CalcOptions = {
    arg1: float
    arg2: float
    operation: CalculatorOperation
}

let isArgLengthSupported (args : string[]) =
    args.Length = 3
    
let parseOperation (arg : string) =
    match arg with
    |"+" -> CalculatorOperation.Plus
    |"-" -> CalculatorOperation.Minus
    |"*" -> CalculatorOperation.Multiply
    |"/" -> CalculatorOperation.Divide
    |_ -> raise (ArgumentException("Wrong operation"))
        
let parseCalcArguments(args : string[]) =
    let mutable n1 = 0
    let mutable n2 = 0
    if not (isArgLengthSupported args) then ArgumentException("Incorrect amount of arguments") |> raise
    if not Double.TryParse args[0] n1 then ArgumentException("First argument must be a number") |> raise
    let parsedOperation = parseOperation args[1]
    if not (Double.TryParse(args[2] n2)) then ArgumentException("Third argument must be a number") |> raise
    
    let result = {
        arg1 = n1;
        arg2 = n2;
        operation = parsedOperation}    
    result 
